### PR TITLE
"container-tests" module.

### PR DESCRIPTION
**Pull request by mark-terry**
## PR description
Adds a new "container-tests" module, with a 2 node private network (besu/tessera, GoQuorum/tessera) running.

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/LOG [hunk 1]
```diff
@@ -0,0 +1,8 @@
+=============== Feb 3, 2021 (AEST) ===============
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/LOG:1]: Will review. 👍

File: container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/CURRENT [hunk 2]
```diff
@@ -0,0 +1 @@
+MANIFEST-000000
```
Comment (pinges) [inline container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/CURRENT:1]: Do we have to check in the files in chaindata and lightchaindata?

File: .circleci/config.yml [hunk 3]
```diff
@@ -191,6 +195,19 @@ jobs:
...
       - capture_test_logs
 
+  containerTests:
+    parallelism: 1
+    executor: machine_executor
+    steps:
+      - prepare
+      - attach_workspace:
+          at: ~/project
+      - run:
```
Comment (usmansaleem) [inline .circleci/config.yml:205]: add another "run" here which runs gradlew distDocker on parent directory

File: container-tests/tests/build.gradle [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  testImplementation 'org.awaitility:awaitility'
+  testImplementation 'org.testcontainers:testcontainers'
+  testImplementation 'org.web3j:core'
+  testImplementation 'org.web3j:quorum'
+}
+
+test.enabled = false
+
+task containerTests(type: Test) {
+  description = 'Runs GoQuorum <> Besu container tests.'
```
Comment (usmansaleem) [inline container-tests/tests/build.gradle:32]: this is where you need to specify dependsOn root project's distDocker ... and re-calculate docker image. Here, you can either set system property or environment variable which will be visible to all tests.

File: container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/CURRENT (hunk 2 above)
Comment (mark-terry) [reply container-tests/tests/src/test/resources/testcontainers/goQuorum/qdata/geth/chaindata/CURRENT]: Removed.

File: container-tests/tests/src/test/java/org/hyperledger/besu/tests/container/ContainerTestBase.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            besuContainerGenesisPath,
+            "--network-id",
+            networkId.toString(),
+            "--p2p-port",
+            besuP2pPort.toString(),
+            "--rpc-http-enabled",
+            "--rpc-http-port",
+            besuRpcPort.toString(),
+            "--rpc-http-api",
+            "ADMIN,ETH,NET,WEB3,PRIV,GOQUORUM",
```
Comment (macfarla) [inline container-tests/tests/src/test/java/org/hyperledger/besu/tests/container/ContainerTestBase.java:172]: if we're in goquorum mode we don't need to enable PRIV

**Review by macfarla — COMMENTED** [review]
looks good overall

File: container-tests/tests/src/test/resources/tessera/tesseraConfig.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    {
+      "app":"ThirdParty",
+      "enabled": true,
+      "serverAddress": "http://localhost:9081",
+      "communicationType" : "REST"
+    },
+    {
+      "app":"Q2T",
+      "enabled": true,
+      "serverAddress": "unix:/tmp/test.ipc",
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply container-tests/tests/src/test/resources/tessera/tesseraConfig.json:19]: There's one Tessera config that's being dynamically overridden in the `buildGoQuorumTesseraContainer` and `buildBesuTesseraContainer`. This specific configuration is just the example config file, specific to GoQuorum.

File: container-tests/tests/src/test/resources/genesis.json [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    "eip155Block": 0,
+    "eip150Hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+    "eip158Block": 0,
+    "clique":{
+      "blockperiodseconds":2,
+      "epochlength":30000,
+      "period": 2,
+      "epoch": 30000
+    },
+    "maxCodeSizeConfig": [
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply container-tests/tests/src/test/resources/genesis.json:17]: You mean Besu uses "contractSizeLimit" instead?  Looks like it's not being used, will remove. 👍

**Review by macfarla — APPROVED** [review]
LGTM

**Pull request merged**